enhancements/machine-api/defaulting-validating.md

```diff
@@ -0,0 +1,143 @@
+
+- The serving certs and CA bundle for the webhook server will be managed by the https://github.com/openshift/service-ca-operator via `service.beta.openshift.io/serving-cert-secret-name:` and `"service.beta.openshift.io/inject-cabundle": "true"` annotations.
+
+#### Implementation
+
+Since most of the machine API controllers rely on controller-runtime we can leverage it to define the webhook server as well.
+We could implement `webhook.Defaulter` and `webhook.Validator` interfaces as in https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html.
+
+However we need a more granular level of customisation which is not satisfied by those interfaces. Particularly we need to know the provider we are running on (to run the appropriate defaulting/validation) and the clusterID (to satisfy some defaults).
+To that end we can implement the [Handler interface](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/webhook/admission#Handler) enabling us to infer the values above at the time of instantiating the webhook rather than on each request at runtime.
+
+### Risks and Mitigations
+
+As we introduce stronger opinions on defaults and validation we might break automated consumers that were relying on faulty behaviour.
+New defaults and validation needs to be reviewed carefully for each particular resource.
+
+## Design Details
+
+### Test Plan
+
+Extensive unit testing to cover defaulting and validation business logic.
+e2e testing on the Machine API test suite to validate we reject invalid input.
+
+### Graduation Criteria
+
+
+### Upgrade / Downgrade Strategy
+
+The webhook will start gating upcoming machine API requests as it runs with new payload.
+Existing functional machines should remain untouched. Either because we choose to disable defaulting for update requests or because our defaults should not apply to non-required fields.
+
+### Version Skew Strategy
+
+
+## Implementation History
+
+[AWS implementation](https://github.com/openshift/machine-api-operator/pull/601)
+
+## Drawbacks
+
+
+## Alternatives
+
+- Introduce a new CRD for the `providerSpec`. Externalise it from the machine object. Introduce an `objectRef` in the machine object to reference `providerSpec` resources. This would let the `providerSpec` become a natural extension of the Kubernetes API. Although we still might need a webhook for clever defaulting/validation we'd get for free all the features coming from [structural schemes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190425-structural-openapi.md)
+
+The above is still desirable but more of an orthogonal long term goal. Once we go that route it means API breaking changes will be disruptive for users.
```

Comment:
Another alternative would be to make the parts of the providerSpec first-class parts of the Machine API itself, instead of using a separate CRD.

I see two ways to do that. Either add all of the provider-specific sections to the Machine CRD and have 1 type for all clusters (replacing `providerSpec` with `aws`, `azure`, etc. structs), or have a provider-specific version of the Machine CRD that is installed based on the platform (exposing the content of `providerSpec` without changing the name). The former would require coordinating between provider implementations to have the API defined in a single place, which might not be ideal. It would make it easy to run multiple providers in the same cluster, for cross-cloud or hybrid metal/cloud clusters, although it isn't clear if this is a product goal. The latter approach would also require some level of coordination to ensure the parts of the API that are not provider-specific are consistent. The second approach also would make it harder to run multiple providers on the same cluster, which would make cross-cloud or hybrid metal/cloud clusters more difficult to build.

Either approach would be easier to make forward/backward compatible than a separate CRD since the upgrade/downgrade step could move data between parts of the same resource that could have the same structure as the current providerSpec. Either approach would give all of the kubernetes API integration benefits of having a separate object.

One obvious downside to extending the Machine API directly is that it would continue to take the OpenShift version of that API further away from the direction the upstream version is going. Perhaps that's why this approach wasn't considered?